### PR TITLE
[4.0] Implement time field

### PR DESCRIPTION
**Pull request by VladikB**
Pull Request for Issue #22140  .

### Summary of Changes

I created new files to rectify the missing of input type="time" with a typical time field. Since IE and Safari don't support this input type, on these browsers it goes to input type="text"

### Testing Instructions

a) 1. Go to  administrator\components\com_users\forms. Edit the first field of user.xml file to 

```
name="name"
type="time"
label="COM_USERS_USER_FIELD_NAME_LABEL"
required="true"
validate="time"
size="30"
```
2. Go to Joomla-Backend -> Users ->Manage -> Click on Your User -> try to change it

b) 1. Go to  administrator\components\com_users\forms. Edit the first field of user.xml file to 

```
name="name"
type="time"
label="COM_USERS_USER_FIELD_NAME_LABEL"
required="true"
validate="time"
size="30"
min="08:00"
max="15:00"
```
2. Go to Joomla-Backend -> Users ->Manage -> Click on Your User-> try to change it

c) 1. Go to  administrator\components\com_users\forms. Edit the first field of user.xml file to 

```
name="name"
type="time"
label="COM_USERS_USER_FIELD_NAME_LABEL"
required="true"
validate="time"
size="30"
step="3"
```
2. Go to Joomla-Backend -> Users ->Manage -> Click on Your User-> try to change it

### Expected result

a) The Name field is a time field with hh:mm
b) The Name field is a time field with hh:mm that goes from 08:00 to 15:00
c) The Name field is a time field with hh:mm:ss and seconds can be changed in step of three seconds

### Documentation Changes Required

If this works, I am going to write a documentation since it is a new field type.

### CODE REVIEW

**Comment by roland-d** [comment]
@VladikB Please fix the codestyle issues:

image

**Comment by hans2103** [comment]
I have tested this item ✅ successfully with Chrome browser
It works as described.

File: libraries/src/Form/Field/TimeField.php [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+use Joomla\CMS\Form\FormField;
+use Joomla\CMS\HTML\HTMLHelper;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Uri\Uri;
+
+/**
+ * Form Field class for the Joomla Platform.
+ * Provides a select list of integers with specified first, last and step values.
+ *
+ * @since  1.7.0
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
 * @since  __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ */
+class TimeField extends FormField
+{
+	/**
+	 * The form field type.
+	 *
+	 * @var    string
+	 * @since  1.7.0
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	 */
+	protected $type = 'Time';
+
+	/**
+	 * The allowable minimal value of the field.
+	 *
+	 * @var    integer
+	 * @since  3.2
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	 */
+	protected $min;
+
+	/**
+	 * The allowable maximal value of the field.
+	 *
+	 * @var    integer
+	 * @since  3.2
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	 */
+	protected $max;
+
+	/**
+	 * Steps between different values
+	 *
+	 * @var    integer
+	 * @since  3.2
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	 */
+	protected $step;
+
+	/**
+	 * Input addon before
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]: Looks like this comment can be removed?

File: libraries/src/Form/Field/TimeField.php [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+	 *
+	 * @var    string
+	 * @since  4.0.0
+	 */
+	/**
+	 * Name of the layout being used to render the field
+	 *
+	 * @var    string
+	 * @since  3.7
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since  __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+	/**
+	 * Method to set certain otherwise inaccessible properties of the form field object.
+	 *
+	 * @param   string  $name   The property name for which to set the value.
+	 * @param   mixed   $value  The value of the property.
+	 *
+	 * @return  void
+	 *
+	 * @since   3.2
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	 */
+
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]: please remove this empty line :)

File: libraries/src/Form/Field/TimeField.php [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+	 * @param   \SimpleXMLElement  $element  The SimpleXMLElement object representing the `<field>` tag for the form field object.
+	 * @param   mixed              $value    The form field value to validate.
+	 * @param   string             $group    The field name group control value. This acts as an array container for the field.
+	 *                                       For example if the field has name="foo" and the group value is set to "bar" then the
+	 *                                       full field name would end up being "bar[foo]".
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @see     FormField::setup()
+	 * @since   3.2
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	/**
+	 * Method to get certain otherwise inaccessible properties from the form field object.
+	 *
+	 * @param   string  $name  The property name for which to get the value.
+	 *
+	 * @return  mixed  The property value or null.
+	 *
+	 * @since   3.2
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+		return parent::__get($name);
+	}
+
+	/**
+	 * Method to get the field input markup.
+	 *
+	 * @return  string  The field input markup.
+	 *
+	 * @since   1.7.0
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

File: libraries/src/Form/Field/TimeField.php [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+	{
+		return $this->getRenderer($this->layout)->render($this->getLayoutData());
+	}
+
+	/**
+	 * Method to get the data to be passed to the layout for rendering.
+	 *
+	 * @return  array
+	 *
+	 * @since 3.7
```
Comment (zero-24) [inline libraries/src/Form/Field/TimeField.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Review by zero-24 — CHANGES_REQUESTED** [review]
Some inline coments and most of it are just the since tags 👍

**Comment by coolcat-creations** [comment]
Nice Feature, thank you!!!

**Comment by dgrammatiko** [comment]
Sorry to ruin the party but this cannot be accepted because it's not supported by the minimum browser compatibility for J4: https://caniuse.com/#feat=input-datetime

**Comment by roland-d** [comment]
@dgrammatiko The problem being that a text field is not supported?

**Comment by dgrammatiko** [comment]
@roland-d no the problem is that you cannot rely on the fallback type of text **wihout** some client side validation. Also I think that there should be also a server side validation (input filter).

**Comment by VladikB** [comment]
"Since IE and Safari don't support this input type, on these browsers it goes to input type="text""

**Comment by dgrammatiko** [comment]
> "Since IE and Safari don't support this input type, on these browsers it goes to input type="text""

So I can then type: `This sucks!` and it will be valid time because there is no client side or server side validation. In other words you're proposing something that in it's current form **is broken**. 
Great, ship it!

**Comment by dgrammatiko** [comment]
Basically you need this: https://github.com/Dan503/time-input-polyfill

**Comment by coolcat-creations** [comment]
Would be so great, if we would appreciate new contributions and find ways to make it work instead of shouting at each other and beeing such aggressive.

**Comment by dgrammatiko** [comment]
@coolcat-creations so, explaining that something is not acceptable in the current state is aggressive?

**Comment by coolcat-creations** [comment]
It's not what you say, but how you say it @dgrammatiko 
"This sucks!" and "Great, ship it!" ... We are here to create a great product with each other, not to be mad ;-)

**Comment by brianteeman** [comment]
You should find everything you need to know about the time input and any issues you need to watch out for here https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time

**Comment by VladikB** [comment]
@brianteeman Can You please check the language strings?

**Comment by VladikB** [comment]
@wilsonge Is the added time rule OK like that?

**Comment by brianteeman** [comment]
> @brianteeman Can You please check the language strings?

Do we even need language strings? Don't the browsers have their own native strings for this?

**Comment by VladikB** [comment]
> > @brianteeman Can You please check the language strings?
> 
> Do we even need language strings? Don't the browsers have their own native strings for this?

I have no idea

**Comment by brianteeman** [comment]
create a time field and see what happens when validation is on

**Comment by VladikB** [comment]
The language strings are primary for the browsers that don't support input type=time. So that if You enter a random text or a false time it won't be accepted and gives the message back.
In browsers that support time You just can't input wrong data.

**Comment by brianteeman** [comment]
Sorry if I wasnt clear.

Use a browser which supports time and copy the error messages from there to be used in the other browsers

**Comment by roland-d** [comment]
There is something going sideways here. The language changes are not in relation to what browsers show but these are the message after the server-side validation is done. In case a validation fails, one of these messages are shown.

**Comment by brianteeman** [comment]
OK here you go and sorry for the confusion and please make sure they are the same frontend and admin

```
JLIB_FORM_FIELD_INVALID_MIN_TIME="The time you entered is before the minimum time."
JLIB_FORM_FIELD_INVALID_MAX_TIME="The time you entered is after the maximum time."
JLIB_FORM_FIELD_INVALID_TIME_INPUT="Invalid time format. Please use hh:mm."
JLIB_FORM_FIELD_INVALID_TIME_INPUT_SECONDS="Invalid time format. Please use hh:mm:ss"
```

**Comment by infograf768** [comment]
Tested with the simple example proposed by OP

This is not very useful for language/countries where AM/PM is not used as one has to add these values in the field. One can't enter 13:30 for example, but 01:30 PM
So, not only it is restricted to some browsers but it is also restricted to people familiar with AM/PM

Screen Shot 2019-09-11 at 18 35 16

EDIT: If it based on system locale, then we may be in bad shape when using a different language in the browser or OS (did not test).

See: https://www.timeanddate.com/time/am-and-pm.html

> Where Is the 12-Hour Format Used?
> 
> Most countries around the world today use the 24-hour system. However, the 12-hour format, including am and pm, is officially used in a number of countries, including the United States, Canada (except Québec), Australia, New Zealand, and the Philippines.

The fact that it is changed to a text field in Safari for example let's enter anything in that field. No error. No indication of the real format to use, etc.

File: administrator/language/en-GB/en-GB.lib_joomla.ini [hunk 14]
```diff
@@ -314,6 +314,10 @@ JLIB_FORM_CHANGE_USER="Select User"
 JLIB_FORM_CONTAINS_INVALID_FIELDS="The form cannot be submitted as it's missing required data. <br> Please correct the marked fields and try again."
 JLIB_FORM_ERROR_FIELDS_CATEGORY_ERROR_EXTENSION_EMPTY="Extension attribute is empty in the category field."
 JLIB_FORM_ERROR_FIELDS_GROUPEDLIST_ELEMENT_NAME="Unknown element type: %s"
+JLIB_FORM_FIELD_INVALID_MAX_TIME="The time you entered is after the maximum time."
+JLIB_FORM_FIELD_INVALID_MIN_TIME="The time you entered is before the minimum time."
+JLIB_FORM_FIELD_INVALID_TIME_INPUT="Invalid time format. Please use hh:mm."
+JLIB_FORM_FIELD_INVALID_TIME_INPUT_SECONDS="Invalid time format. Please use hh:mm:ss"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.lib_joomla.ini]:
```suggestion
JLIB_FORM_FIELD_INVALID_TIME_INPUT_SECONDS="Invalid time format. Please use hh:mm:ss."
```

File: language/en-GB/en-GB.lib_joomla.ini [hunk 15]
```diff
@@ -314,6 +314,10 @@ JLIB_FORM_ERROR_NO_DATA="No data."
 JLIB_FORM_ERROR_VALIDATE_FIELD="Invalid xml field."
 JLIB_FORM_ERROR_XML_FILE_DID_NOT_LOAD="XML file did not load."
 JLIB_FORM_FIELD_INVALID="Invalid field:&#160"
+JLIB_FORM_FIELD_INVALID_MAX_TIME="The time you entered is after the maximum time."
+JLIB_FORM_FIELD_INVALID_MIN_TIME="The time you entered is before the minimum time."
+JLIB_FORM_FIELD_INVALID_TIME_INPUT="Invalid time format. Please use hh:mm."
+JLIB_FORM_FIELD_INVALID_TIME_INPUT_SECONDS="Invalid time format. Please use hh:mm:ss"
```
Comment (brianteeman) [inline language/en-GB/en-GB.lib_joomla.ini]:
```suggestion
JLIB_FORM_FIELD_INVALID_TIME_INPUT_SECONDS="Invalid time format. Please use hh:mm:ss."
```

File: layouts/joomla/form/field/time.php [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+ * @var   string   $value           Value attribute of the field.
+ * @var   array    $checkedOptions  Options that will be set as checked.
+ * @var   boolean  $hasValue        Has this field a value assigned?
+ * @var   array    $options         Options available for this field.
+ * @var   array    $inputType       Options available for this field.
+ * @var   array    $spellcheck      Options available for this field.
+ * @var   string   $accept          File types that are accepted.
+ */
+
+$attributes = array(
```
Comment (Quy) [inline layouts/joomla/form/field/time.php]:
```suggestion
$attributes = [
```

File: layouts/joomla/form/field/time.php [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    $disabled ? 'disabled' : '',
+    $readonly ? 'readonly' : '',
+    strlen($hint) ? 'placeholder="' . htmlspecialchars($hint, ENT_COMPAT, 'UTF-8') . '"' : '',
+    !empty($onchange) ? 'onchange="' . $onchange . '"' : '',
+    isset($max) ? 'max="' . $max . '"' : '',
+    isset($step) ? 'step="' . $step . '"' : '',
+    isset($min) ? 'min="' . $min . '"' : '',
+    $required ? 'required' : '',
+    $autofocus ? 'autofocus' : ''
+);
```
Comment (Quy) [inline layouts/joomla/form/field/time.php]:
```suggestion
];
```

**Comment by brianteeman** [comment]
> So, not only it is restricted to some browsers but it is also restricted to people familiar with AM/PM

The browsers take the time format from the operating system locale

> The fact that it is changed to a text field in Safari for example let's enter anything in that field. No error. No indication of the real format to use, etc.

Thats easily resolved by adding a pattern in the xml of the field definition. The time input wont use it but the fallback text input will

>            pattern="[0-9]{2}:[0-9]{2}">

https://codepen.io/brianteeman/pen/xxKjPgz

**Comment by roland-d** [comment]
> No error. No indication of the real format to use, etc.
@infograf768 The test instructions were missing the latest change, so I updated them. Add the `validate="time"` and you will get error messages depending on your error.

**Comment by infograf768** [comment]
> The browsers take the time format from the operating system locale

Not so obvious. Here on Macintosh, these are my settings:

Screen Shot 2019-09-12 at 08 26 51

I checked my system locale here.  It is
```
imac-2:~ mac$ defaults read NSGlobalDomain AppleLocale
en_FR
```

Therefore I should get a 24 h for the time field.
It simply does not.

**Comment by roland-d** [comment]
@infograf768 Which browser are you using? I want to see what I get here.

**Comment by infograf768** [comment]
I found the only way to get a 24h time field
I have to set up my Mac as (French first):
Screen Shot 2019-09-12 at 08 52 36

Which I don't like as I am used to an English Interface for more than 30 years...

Note: I use Firefox, but it is unrelated.

**Comment by roland-d** [comment]
These are my settings, being Dutch I like it in English as well.
image

This is what I see in Firefox:
image

**Comment by infograf768** [comment]
Lucky you... I use Sierra 12.10.6

Concerning validation. I added validate="time" in the xml for user name.
I do  get a warning indeed in Safari, but it flashes so fast I can't even read it.

error_timefield

**Comment by roland-d** [comment]
@infograf768 That flashing message is unrelated to this PR. The message is just added the message queue of Joomla.

However for me nothing flashes, all Joomla generated messages stay on screen until I close the message.

**Comment by VladikB** [comment]
Time fieldt


https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time

**Comment by infograf768** [comment]
@roland-d 
> However for me nothing flashes, all Joomla generated messages stay on screen until I close the message.

Lucky you. I tested other messages in Safari and they do stay on screen until I close, but not this one.

**Comment by infograf768** [comment]
@VladikB 
I understand how it behaves in Firefox.
My issue here is with the locale.

**Comment by roland-d** [comment]
@infograf768 I was testing in Firefox. Doing the same test in Safari, there the message flashes indeed. However it doesn't fully disappear:
image

There are more issues in Safari. When I go to the System dashboard it loads like this:
image

a second later it looks like this:
image

I would say we move the Safari rendering to another discussion as this is not related to the time field.

As for the locale issue, I don't see how we can do something about that because it is browser that acts upon the time input field. To rule anything Joomla out, what shows up if you create a plain html file with just this field? Attached is a zip file with the HTML file with just the input.

[time.zip](https://github.com/joomla/joomla-cms/files/3604224/time.zip)

**Comment by brianteeman** [comment]
Or test with the codepen I posted yesterday https://codepen.io/brianteeman/pen/xxKjPgz

**Comment by infograf768** [comment]
@roland-d 
Joomla is not at stake.
I still get the field with AM/PM with your html in Firefox as it really depends on my Mac settings.

Concerning Safari, I confirm your findings for some of the quickicons modules and agree it may have to be solved separately.

@brianteeman 
Looks like your solution would help entering something which makes sense in the field in Safari, forcing 24h. Evidently the Label needs to be very informative, which may be confusing when the browser displays the time field as one, for example, would not be able to enter `17:00` in your proposal, but `05:00 PM`

Screen Shot 2019-09-12 at 10 10 29


Note: When one gets the AM/PM field display in Firefox it shows like this:

Screen Shot 2019-09-12 at 10 06 37

It is not obvious that the last 2 `--` mean entering AM or PM

**Comment by roland-d** [comment]
The problem of all the dashes can simply be dealt with by providing a default value.

**Comment by infograf768** [comment]
> The problem of all the dashes can simply be dealt with by providing a default value.

Would not you need one when the time field is using AM/PM, one when using 24h + one when the time field is a text field?   ;)

**Comment by roland-d** [comment]
I think you only need 1 default value.

**Comment by wilsonge** [comment]
This will do as a starting point. I don't think we need the polyfill and the validation looks good. As this is a browser native field I don't think there is much we can be doing to work around the differences - even if there are issues. We can consider trying to normalise in the future i guess. But i don't see this being much different to things like when we added the number field etc way back.

**Comment by wilsonge** [comment]
@infograf768 i agree btw that before this is used in core we need to evaluate the lack of consistency. However I think allowing people to use native browser fields is a good idea

**Pull request merged**